README.md
[![npm](https://img.shields.io/npm/v/aria-modal.svg)](https://npmjs.org/package/aria-modal) [![Published on webcomponents.org](https://img.shields.io/badge/webcomponents.org-published-blue.svg)](https://www.webcomponents.org/element/aria-modal)

# aria-modal

Accessible modal with Web Components

## About

aria-modal that is a fully accessible is built according to [WAI-ARIA Authoring Practices](https://www.w3.org/TR/wai-aria-practices/#dialog_modal). And it provide the only simple features. It makes be easy to implement accessible modal. If you use polyfill, you can use [@webcomponents/webcomponentsjs](https://github.com/webcomponents/polyfills/tree/master/packages/webcomponentsjs).

## Features

- Focus is moved to first-focus element. Tab and Shift+Tab keys will cycle through the modal's focusable nodes.
- ESC key closes the modal.
- Scrolling is frozen on the document.body.
- When the modal closes, focus returns to the element that was focused before the modal are shown.
- The dialog element has an ARIA role of dialog by default(You can change role).
- The dialog element must has either aria-label or aria-labelledby.
- If clicking the backdrop, the modal is closed.
- If animation property is true, the modal can use the fade-in(out) animation.

## DEMO

https://keiya01.github.io/aria-modal-doc/top/

## Install

Using npm:

```bash
$ npm install aria-modal
```

## Usage

```js

import 'aria-modal';

/**
 * do something
*/

```

### Append Your Modal

You can append your modal inside aria-modal with `slot`. If you use `slot`, you can use accessible feature for modal.  
Please see following example.

```html

<aria-modal
  app="app"
  first-focus="button"
  animation="true"
  active="active"
>
  <div id="node" slot="modal" role="dialog" aria-labelledby="title">
    <!-- ... -->
    <button id="button"></button>
  </div>
</aria-modal>

```

If you use custom element, see the following code.

```html

<aria-modal
  app="app"
  active="active"
  animation="true"
>
  <!-- 
    <simple-modal> must be contained `firstFocus` function that return HTMLElement.
    And You must set role and aria-* attributes to your modal.
  -->
  <simple-modal slot="modal"></simple-modal>
</aria-modal>

```

### Style

Using css variables, you can apply your style to `<aria-modal>`.

```css

aria-modal {
  --backdrop-display: block; /* or flex, inline-block, etc... */
  --backdrop-color: rgba(0, 0, 0, 0.3); /* background-color for backdrop */
  --backdrop-position: fixed; /* or fixed */
  --backdrop-z-index: none;
  --animation-function: linear;
  --animation-duration: 300ms;
  /* required, if fade property is false */
  --animation-delay: none;
}

```

#### Tips

If you want to set your modal to center, you can use `margin: auto;` and `flexbox` property.

```css

aria-modal {
  --backdrop-display: flex;
  /* ... */
}

.your-modal {
  margin: auto;
  height: 300px;
  width: 500px;
  background-color: #fff;
}

```

### Props

#### open 

Required: `false`, Type: `boolean`, Default: `false`  
  
It is used to show modal. You can set `true` if you want to open modal.

#### first-focus

Required: `true`, Type: `string` or `function firstFocus(): HTMLElement`  
  
It is used to focus to first element when modal is opened. You should assign `id` name.  
  
If `first-focus` element is a custom element or inside a custom element, You must implement `firstFocus` method to your slotted element. That is, slotted element must be custom element if `first-focus` element is custom element.

```js

class SampleModal extends HTMLElement {
  get template() {
    const template = document.createElement('template');
    template.innerHTML = `
      <div>
        <button id="first-element">Button</button>
      </div>
    `;
  }
  
  constructor() {
    super();

    const shadowRoot = this.attachShadow({ mode: 'open' });
    shadowRoot.appendChild(this.template.content.cloneNode(true));
  }

  /**
   * Do something
  */

 // You must implement `firstFocus()`, if you use custom element.
 firstFocus() {
   return this.shadowRoot.getElementById('first-element');
 }
}

```

#### app

Required: `true`, Type: `string`  
  
It is used to set `aria-hidden` to your `app` element. You should set main contents `id` name.
  
**Notice**:  Don't contain `aria-modal` inside main contents. If you contain `aria-modal` to your main contents, screen readers can not find `aria-modal`.

#### animation

Required: `false`, Type: `boolean`, Default: `false`  
  
If `true`, you can use animation. aria-modal run fade animation by default.
If you want to use custom animation, you can use `active` and `hide` properties.

#### fade

Required: `false`, Type: `boolean`, Default: `true`

If `false`, default fade animation is disabled.

#### active

Required: `false`, Type: `string`  
  
`active` is class name that is added when `open` props is changed to `true`.

#### hide

Required: `false`, Type: `string`  
  
`hide` is class name that is added when `open` props is changed to `false`.

#### disabled

Required: `false`, Type: `boolean`  
  
If true, disable clicking on the backdrop.

#### aria-label or aria-labelledby

Required: `true`, Type: `string`  
  
You must include `aria-label` or `aria-labelledby` property to your modal.  
See https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/ARIA_Techniques/Using_the_aria-labelledby_attribute

#### aria-describedby

Required: `false`, Type: `string`  
  
See https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/ARIA_Techniques/Using_the_aria-describedby_attribute

#### role

Required: `true`, Type: `string`  
  
You must include `role` property to your modal.  
Assignable value is `dialog` or `alertdialog`. See https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/Roles

#### aria-modal

Required: `false`, Type: `boolean`, Default: `false`  
  
This property is some problems, see this article https://developer.paciellogroup.com/blog/2018/06/the-current-state-of-modal-dialog-accessibility/  
And if you want to know about `aria-modal`, see https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/Roles/dialog_role

### Warning

If the active class is not set to `aria-modal`, it may flicker when reloaded on browser. When you set active class, you should set your modal to `display: none;` and you should set your active class to `display: block; /* or flex, inline-block, etc... */`.  
This problem is occurred by rendering process for Web Components. Your modal is assigned to `slot` element in `aria-modal`, and `slot` is rendered after JavaScript file(and Web Components) have loaded. That is, this problem is occurred by your modal is rendered without your modal is assigned to slot.

### Example

https://github.com/keiya01/aria-modal/tree/master/example

## License

[MIT License](https://github.com/keiya01/aria-modal/blob/master/LICENSE)


CHANGELOG.md
# Changelog

## v2.6.9

- Removed `node` property
- Removed `shadow` property
- Fixed bugs

## v2.6.8

- Fixed bugs

## v2.6.7

- Fixed bugs

## v2.6.6

- Fixed bugs
- Added `disabled` property that is not to prevent hide when clicked backdrop.

## v2.6.5

- Fixed bugs

## v2.6.4

- Fixed README.ms

## v2.6.3

- Added `fade` property to opt out  default fade animation.
- Added `--animation-delay` css variables to wait finishing your custom animation.
- Added fallback to css variables.

## v2.6.2

- Fixed css variables description on README.md

## v2.6.1

- Fixed a bug that dialog does not open when `open` property is `true` at initial loading.

## v2.6.0

- **Breaking Change**: Deleted `duration` property and added `--animation-duration` CSS variables.
- Fixed README.md
- Fixed performance
- Reduced code size
- Fixed bugs

## v2.5.1

- Fixed README.md

## v2.5.0

- **Breaking Change**: Added `app` property. This property is require to run aria-modal.
- **Breaking Change**: Changed default value of `aria-modal` property to `false`, because `aria-modal` property is contained some bugs.
- Fixed to set default `display` style to `display: block;` for aria-modal(`:host`)

## v2.4.2, v2.4.3, v2.4.4

- Fixed bugs that web contents is not read because aria-modal is **visible**.

## v2.4.1

- Fixed bugs

## v2.4.0

- Added `hide` property to run animation for hide phase
- Deleted `aria-*` and `role` validation
- Fixed bugs

## v2.3.1

- Fixed bugs that the modal don't open when initial `open` property is true

## v2.3.0

- Fixed types on index.d.ts

## v2.2.5

- Supported scrolling on the modal

## v2.2.4

- Fixed bugs

## v2.2.3

- Publish DEMO page

## v2.2.1, v2.2.2

- Fixed README.md

## v2.2.0

- Supported frozen scroll

## v2.1.1

- Fixed bugs

## v2.1.0

- Refactor source code
- Deleted `AriaModalProps`
- Fixed bugs

## v2.0.4

- Fixed bugs

## v2.0.1, v2.0.2, v2.0.3

- Fixed README.md

## v2.0.0

- Supported custom element.
- Added `shadow` property that is used to check your node element is custom element.
- Fixed `firstFocus` property.
  - `firstFocus` property is supported custom element.
  - You can use custom property, if you implement `firstFocus` function to your node element instead of `firstFocus` property.
- Changed `aria-describedby` to optional.

## v1.2.10

- Fixed README.md

## v1.2.9

- Fixed LICENSE
- Fixed README.md

## v1.2.8

- Fixed README.md

## v1.2.7

- Fixed README.md

## v1.2.6

- Fixed bugs

## v1.2.5

- Fixed bugs

## v1.2.4

- Fixed README.md

## v1.2.3

- **Breaking Change**: Renamed package name from web-components-aria-modal to aria-modal.

## v1.2.2

- Added `--backdrop-z-index` as CSS variables.

## v1.2.1

- Minified UMD file
- **Breaking Change**: Fixed type in index.d.ts

## v1.2.0

- Added `active` property.
  - `active` property is added to modal class list, if animation property is true.
- Added `--backdrop-display` and `--animation-function` as CSS variables.

## v1.1.0

- Added `animation` property.
  - Run fade in animation, if `animation` property is true.
- Added `duration` property.
  - animation duration.
- **Breaking Changes**: Delete modal default styles.
- **Breaking Changes**: CSS variables for modal.

## v1.0.0

- first release.


src/aria-modal.test.ts
it('sample', () => {
  expect('test').toBe('test');
});

jest.config.js
// For a detailed explanation regarding each configuration property, visit:
// https://jestjs.io/docs/en/configuration.html

module.exports = {
  // All imported modules in your tests should be mocked automatically
  // automock: false,

  // Stop running tests after `n` failures
  // bail: 0,

  // Respect "browser" field in package.json when resolving modules
  // browser: false,

  // The directory where Jest should store its cached dependency information
  // cacheDirectory: "/private/var/folders/19/3480x7x51wb1b21md2v2lw_h0000gn/T/jest_dx",

  // Automatically clear mock calls and instances between every test
  clearMocks: true,

  // Indicates whether the coverage information should be collected while executing the test
  // collectCoverage: false,

  // An array of glob patterns indicating a set of files for which coverage information should be collected
  // collectCoverageFrom: undefined,

  // The directory where Jest should output its coverage files
  // coverageDirectory: undefined,

  // An array of regexp pattern strings used to skip coverage collection
  // coveragePathIgnorePatterns: [
  //   "/node_modules/"
  // ],

  // A list of reporter names that Jest uses when writing coverage reports
  // coverageReporters: [
  //   "json",
  //   "text",
  //   "lcov",
  //   "clover"
  // ],

  // An object that configures minimum threshold enforcement for coverage results
  // coverageThreshold: undefined,

  // A path to a custom dependency extractor
  // dependencyExtractor: undefined,

  // Make calling deprecated APIs throw helpful error messages
  // errorOnDeprecated: false,

  // Force coverage collection from ignored files using an array of glob patterns
  // forceCoverageMatch: [],

  // A path to a module which exports an async function that is triggered once before all test suites
  // globalSetup: undefined,

  // A path to a module which exports an async function that is triggered once after all test suites
  // globalTeardown: undefined,

  // A set of global variables that need to be available in all test environments
  // globals: {},

  // The maximum amount of workers used to run your tests. Can be specified as % or a number. E.g. maxWorkers: 10% will use 10% of your CPU amount + 1 as the maximum worker number. maxWorkers: 2 will use a maximum of 2 workers.
  // maxWorkers: "50%",

  // An array of directory names to be searched recursively up from the requiring module's location
  // moduleDirectories: [
  //   "node_modules"
  // ],

  // An array of file extensions your modules use
  // moduleFileExtensions: [
  //   "js",
  //   "json",
  //   "jsx",
  //   "ts",
  //   "tsx",
  //   "node"
  // ],

  // A map from regular expressions to module names that allow to stub out resources with a single module
  // moduleNameMapper: {},

  // An array of regexp pattern strings, matched against all module paths before considered 'visible' to the module loader
  // modulePathIgnorePatterns: [],

  // Activates notifications for test results
  // notify: false,

  // An enum that specifies notification mode. Requires { notify: true }
  // notifyMode: "failure-change",

  // A preset that is used as a base for Jest's configuration
  // preset: undefined,

  // Run tests from one or more projects
  // projects: undefined,

  // Use this configuration option to add custom reporters to Jest
  // reporters: undefined,

  // Automatically reset mock state between every test
  // resetMocks: false,

  // Reset the module registry before running each individual test
  // resetModules: false,

  // A path to a custom resolver
  // resolver: undefined,

  // Automatically restore mock state between every test
  // restoreMocks: false,

  // The root directory that Jest should scan for tests and modules within
  // rootDir: undefined,

  // A list of paths to directories that Jest should use to search for files in
  // roots: [
  //   "<rootDir>"
  // ],

  // Allows you to use a custom runner instead of Jest's default test runner
  // runner: "jest-runner",

  // The paths to modules that run some code to configure or set up the testing environment before each test
  // setupFiles: [],

  // A list of paths to modules that run some code to configure or set up the testing framework before each test
  // setupFilesAfterEnv: [],

  // A list of paths to snapshot serializer modules Jest should use for snapshot testing
  // snapshotSerializers: [],

  // The test environment that will be used for testing
  testEnvironment: "jsdom",

  // Options that will be passed to the testEnvironment
  // testEnvironmentOptions: {},

  // Adds a location field to test results
  // testLocationInResults: false,

  // The glob patterns Jest uses to detect test files
  // testMatch: [
  //   "**/__tests__/**/*.[jt]s?(x)",
  //   "**/?(*.)+(spec|test).[tj]s?(x)"
  // ],

  // An array of regexp pattern strings that are matched against all test paths, matched tests are skipped
  // testPathIgnorePatterns: [
  //   "/node_modules/"
  // ],

  // The regexp pattern or array of patterns that Jest uses to detect test files
  // testRegex: [],

  // This option allows the use of a custom results processor
  // testResultsProcessor: undefined,

  // This option allows use of a custom test runner
  // testRunner: "jasmine2",

  // This option sets the URL for the jsdom environment. It is reflected in properties such as location.href
  // testURL: "http://localhost",

  // Setting this value to "fake" allows the use of fake timers for functions such as "setTimeout"
  // timers: "real",

  // A map from regular expressions to paths to transformers
  // transform: undefined,

  // An array of regexp pattern strings that are matched against all source file paths, matched files will skip transformation
  // transformIgnorePatterns: [
  //   "/node_modules/"
  // ],

  // An array of regexp pattern strings that are matched against all modules before the module loader will automatically return a mock for them
  // unmockedModulePathPatterns: undefined,

  // Indicates whether each individual test should be reported during the run
  // verbose: undefined,

  // An array of regexp patterns that are matched against all source file paths before re-running tests in watch mode
  // watchPathIgnorePatterns: [],

  // Whether to use watchman for file crawling
  // watchman: true,
};


example/index.umd.js
(function (global, factory) {
  typeof exports === 'object' && typeof module !== 'undefined' ? module.exports = factory() :
  typeof define === 'function' && define.amd ? define(factory) :
  (global = global || self, global.AriaModalElement = factory());
}(this, (function () { 'use strict';

  const roles = ['dialog', 'alertdialog'];
  class AriaModalElement extends HTMLElement {
      constructor() {
          var _a, _b;
          super();
          this.moveFocusToFirst = (e) => {
              const target = e.target;
              this.focusFirstElement(target, this.props.node);
          };
          this.moveFocusToLast = (e) => {
              const target = e.target;
              this.focusLastElement(target, this.props.node);
          };
          this.handleOnKeyup = (e) => {
              const key = e.keyCode;
              if (key === 27 && this.props.open) {
                  this.setAttribute('open', 'false');
                  e.stopPropagation();
              }
          };
          this.focusAfterClose = null;
          this.props = {
              open: this.getAttribute('open') === 'true',
              node: this.getElementByAttribute('node'),
              firstFocus: this.getElementByAttribute('first-focus'),
              animation: this.getAttribute('animation') === 'true',
              duration: Number(this.getAttribute('duration')) || 300,
          };
          this.validateAriaAttrs(['aria-label', 'aria-labelledby']);
          this.validateAriaAttrs(['aria-describedby']);
          const role = this.getAttribute('role') || 'dialog';
          if (!roles.includes(role)) {
              throw new Error(`role attribution is assigned invalid value. assignable value are ${roles.join(' or ')}.`);
          }
          const ariaModal = this.getAttribute('aria-modal');
          if (!ariaModal) {
              this.setAttribute('aria-modal', 'true');
          }
          const shadowRoot = this.attachShadow({ mode: 'open' });
          shadowRoot.appendChild(this.template().content.cloneNode(true));
          document.addEventListener('keyup', this.handleOnKeyup);
          (_a = shadowRoot.getElementById('first-descendant')) === null || _a === void 0 ? void 0 : _a.addEventListener('focus', this.moveFocusToLast, true);
          (_b = shadowRoot.getElementById('last-descendant')) === null || _b === void 0 ? void 0 : _b.addEventListener('focus', this.moveFocusToFirst, true);
      }
      static get observedAttributes() {
          return ['open'];
      }
      attributeChangedCallback(name, _, newValue) {
          if (name === 'open') {
              this.props.open = newValue === 'true';
              this.changeStyle();
              this.setTabIndex();
              this.trapFocus();
          }
      }
      disconnectedCallback() {
          var _a, _b, _c, _d;
          (_b = (_a = this.shadowRoot) === null || _a === void 0 ? void 0 : _a.getElementById('first-descendant')) === null || _b === void 0 ? void 0 : _b.addEventListener('focus', this.moveFocusToLast, true);
          (_d = (_c = this.shadowRoot) === null || _c === void 0 ? void 0 : _c.getElementById('last-descendant')) === null || _d === void 0 ? void 0 : _d.addEventListener('focus', this.moveFocusToFirst, true);
      }
      adoptedCallback() {
          var _a, _b, _c, _d;
          (_b = (_a = this.shadowRoot) === null || _a === void 0 ? void 0 : _a.getElementById('first-descendant')) === null || _b === void 0 ? void 0 : _b.addEventListener('focus', this.moveFocusToLast, true);
          (_d = (_c = this.shadowRoot) === null || _c === void 0 ? void 0 : _c.getElementById('last-descendant')) === null || _d === void 0 ? void 0 : _d.addEventListener('focus', this.moveFocusToFirst, true);
      }
      trapFocus() {
          var _a;
          if (this.props.open) {
              this.focusAfterClose = document.activeElement;
              this.props.firstFocus.focus();
          }
          else {
              (_a = this.focusAfterClose) === null || _a === void 0 ? void 0 : _a.focus();
          }
      }
      setTabIndex() {
          var _a;
          const modal = (_a = this.shadowRoot) === null || _a === void 0 ? void 0 : _a.querySelector('slot');
          const prevSibling = modal === null || modal === void 0 ? void 0 : modal.previousElementSibling;
          const nextSibling = modal === null || modal === void 0 ? void 0 : modal.nextElementSibling;
          if (prevSibling) {
              prevSibling.setAttribute('tabindex', '0');
          }
          if (nextSibling) {
              nextSibling.setAttribute('tabindex', '0');
          }
      }
      setHideStyle(backdrop) {
          if (this.props.animation) {
              backdrop.classList.add('hide');
              setTimeout(() => {
                  backdrop.classList.remove('active');
                  backdrop.classList.remove('hide');
              }, this.props.duration);
          }
          else {
              backdrop.classList.remove('active');
          }
      }
      changeStyle() {
          var _a;
          const backdrop = (_a = this.shadowRoot) === null || _a === void 0 ? void 0 : _a.getElementById("aria-modal-backdrop");
          if (!backdrop) {
              throw new Error('Could not find aria-modal-backdrop id');
          }
          if (this.props.open) {
              backdrop.classList.add('active');
          }
          else {
              this.setHideStyle(backdrop);
          }
      }
      template() {
          const template = document.createElement('template');
          document.body.appendChild(template);
          template.innerHTML = `
      <style>
        @keyframes fade-in {
          from {
            opacity: 0;
          }
          to {
            opacity: 1;
          }
        }
        @keyframes fade-out {
          from {
            opacity: 1;
          }
          to {
            opacity: 0;
          }
        }
        .backdrop {
          display: none;
          background-color: var(--backdrop-color, rgba(0, 0, 0, 0.6));
          position: var(--backdrop-position, absolute);
          top: 0;
          right: 0;
          bottom: 0;
          left: 0;
          ${this.props.animation ? 'opacity: 0;' : ''}
        }
        .backdrop.active {
          display: var(--backdrop-display, block);
          ${this.props.animation
            ?
                `animation: fade-in ${this.props.duration}ms var(--animation-function, ease-in) forwards;`
            :
                ''}
        }
        .backdrop.hide {
          ${this.props.animation
            ?
                `animation: fade-out ${this.props.duration}ms var(--animation-function, ease-in) forwards;`
            :
                ''}
        }
      </style>
      <div id="aria-modal-backdrop" class="backdrop">
        <div id="first-descendant"></div>
          <slot name="modal"></slot>
        <div id="last-descendant"></div>
      </div>
    `;
          return template;
      }
      validateAriaAttrs(arr) {
          const validArr = [];
          arr.map(val => {
              if (this.getAttribute(val)) {
                  validArr.push(val);
              }
          });
          if (validArr.length === 0) {
              throw new Error(`${arr.join(' or ')} must be included on aria-modal.`);
          }
          if (validArr.length >= 2) {
              throw new Error(`${arr.join(' or ')} can include just one on aria-modal.`);
          }
          return validArr[0];
      }
      getElementByAttribute(name) {
          const id = this.getAttribute(name);
          if (!id) {
              throw new Error(`${name} is not assigned`);
          }
          const element = document.getElementById(id);
          if (!element) {
              throw new Error(`${name} could not find. first-focus must be assigned id name.`);
          }
          return element;
      }
      isFocusable(target, element) {
          return document.activeElement !== target && document.activeElement === element;
      }
      focusFirstElement(target, node) {
          const children = node.children;
          for (let i = 0; i < children.length; i++) {
              const child = children[i];
              child.focus();
              if (this.isFocusable(target, child) || this.focusFirstElement(target, child)) {
                  return true;
              }
          }
          return false;
      }
      focusLastElement(target, node) {
          const children = node.children;
          for (let i = children.length - 1; i >= 0; i--) {
              const child = children[i];
              child.focus();
              if (this.isFocusable(target, child) || this.focusLastElement(target, child)) {
                  return true;
              }
              return false;
          }
      }
  }

  if (!window.customElements.get('aria-modal')) {
      window.AriaModalElement = AriaModalElement;
      window.customElements.define('aria-modal', AriaModalElement);
  }

  return AriaModalElement;

})));


src/aria-modal.ts
type ModalNode = HTMLElement & { firstFocus?: () => HTMLElement };
export default class AriaModalElement extends HTMLElement {
  private focusAfterClose: HTMLElement | null;
  private loaded?: boolean;

  get styles() {
    return `
      <style>
        :host {
          display: var(--backdrop-display, block);
          background-color: var(--backdrop-color, rgba(0, 0, 0, 0.3));
          position: var(--backdrop-position, fixed);
          top: 0;
          right: 0;
          bottom: 0;
          left: 0;
          z-index: var(--backdrop-z-index);
          overflow-y: auto;
          opacity: 0;
        }

        :host([animation="true"]) {
          will-change: opacity;
        }

        :host([open="true"]) {
          opacity: 1;
        }
      </style>
    `;
  }

  get template() {
    const template = document.createElement('template');

    template.innerHTML = `
      ${this.styles}
      <div id="first-descendant"></div>
      <slot name="modal"></slot>
      <div id="last-descendant"></div>
    `;
  
    return template;
  }

  constructor() {
    super();
    
    this.focusAfterClose = null;

    const shadowRoot = this.attachShadow({ mode: 'open' });
    shadowRoot.appendChild(this.template.content.cloneNode(true));
  }

  static get observedAttributes() {
    return ['open'];
  }

  attributeChangedCallback(name: string) {
    if(name === 'open' && this.loaded) {
      return this.handleOnOpen();
    }
    this.loaded = true;
  }

  connectedCallback() {
    if(this.open) {
      window.addEventListener('DOMContentLoaded', this.handleOnDOMContentLoaded);
    } else {
      this.open = false;
      this.style.visibility = 'hidden';
    }

    document.addEventListener('keyup', this.handleOnKeyup);
    this.addEventListener('click', this.handleOnClickBackdrop);
    this.addEventListener('transitionend', this.handleOnTransitionEnd);
    this.shadowRoot!.getElementById('first-descendant')?.addEventListener('focus', this.moveFocusToLast, true);
    this.shadowRoot!.getElementById('last-descendant')?.addEventListener('focus', this.moveFocusToFirst, true);
  }
  
  disconnectedCallback() {
    if(this.open) {
      window.removeEventListener('DOMContentLoaded', this.handleOnDOMContentLoaded);
    }
    document.removeEventListener('keyup', this.handleOnKeyup);
    this.removeEventListener('click', this.handleOnClickBackdrop);
    this.removeEventListener('transitionend', this.handleOnTransitionEnd);
    this.shadowRoot?.getElementById('first-descendant')?.removeEventListener('focus', this.moveFocusToLast, true);
    this.shadowRoot?.getElementById('last-descendant')?.removeEventListener('focus', this.moveFocusToFirst, true);
  }
  
  get open() {
    return this.getAttribute('open') === 'true';
  }

  set open(newValue: boolean) {
    this.setAttribute('open', `${newValue}`);
  }

  get app() {
    return this.getElementByAttribute('app');
  }

  get node() {
    const elm = this.shadowRoot!.querySelector('slot')!.assignedElements()[0];
    if(!elm) {
      throw new Error('Could not find element to be specified to slot element.');
    }
    return elm as ModalNode;
  }

  get animation() {
    return this.getAttribute('animation') === 'true';
  }

  get active() {
    return this.getAttribute('active') || '';
  }

  get hide() {
    return this.getAttribute('hide') || '';
  }

  get fade() {
    return this.getAttribute('fade') !== 'false';
  }

  get disabled() {
    return this.getAttribute('disabled') === 'true';
  }

  get shadowNode() {
    if(!this.node.shadowRoot) {
      throw new Error('node property is not custom element.');
    }
    const child = this.node.shadowRoot.children.item(0);
    if(!child) {
      throw new Error('Element that is specified as slot can contain only 1 child element.');
    }
    return child as ModalNode;
  }

  private getElementByAttribute(name: string) {
    const id = this.getAttribute(name);
    if(!id) {
      throw new Error(`${name} is not assigned`);
    }
    const element = document.getElementById(id);
    if(!element) {
      throw new Error(`${name} could not find. ${name} must be assigned id name.`);
    }
    return element;
  }

  private getActiveElement(target: HTMLElement) {
    let result = target;
    if(target.shadowRoot) {
      result = this.getActiveElement(target.shadowRoot.activeElement as HTMLElement);
    }
    return result;
  }

  private focusFirst = () => {
    this.focusAfterClose = document.activeElement as HTMLElement;
    
    if(this.focusAfterClose.shadowRoot) {
      this.focusAfterClose = this.getActiveElement(this.focusAfterClose.shadowRoot!.activeElement as HTMLElement);
    }
    let firstFocus: HTMLElement | null = null;
    if(this.node.shadowRoot) {
      if(!this.node.firstFocus) {
        throw new Error('firstFocus function could not find. If you use shadow dom, please define firstFocus function.')
      }
      firstFocus = this.node.firstFocus();
    } else {
      firstFocus = this.getElementByAttribute('first-focus');
    }
    setTimeout(() => firstFocus!.focus());
  }

  // Schedule the focusBack method to run after all process have been finished.
  private focusBack() {
    setTimeout(() => {
      if(this.focusAfterClose) {
        setTimeout(() => this.focusAfterClose!.focus());
      }
    });
  }

  private changeModalClassList(method: 'add' | 'remove', props: 'active' | 'hide') {
    if(!this[props]) {
      return;
    }
    
    if(this.node.shadowRoot) {
      this.shadowNode.classList[method](this[props]);
    } else {
      this.node.classList[method](this[props]);
    }
  }

  private setActiveStyle() {
    document.body.style.overflow = 'hidden';
    this.changeModalClassList('add', 'active');
  }
  
  private setHideStyle() {
    document.body.style.overflow = 'auto';
    if(this.animation) {
      this.changeModalClassList('add', 'hide');
    } else {
      this.changeModalClassList('remove', 'active');
      this.style.visibility = 'hidden';
    }
  }

  private setTabIndex() {
    const modal = this.shadowRoot?.querySelector('slot');
    const prevSibling = modal?.previousElementSibling;
    const nextSibling = modal?.nextElementSibling;
    if(!prevSibling || !nextSibling) {
      return;
    }
    if(this.open) {
      prevSibling.setAttribute('tabindex', '0');
      nextSibling.setAttribute('tabindex', '0');
    } else {
      prevSibling.removeAttribute('tabindex');
      nextSibling.removeAttribute('tabindex');
    }
  }

  private handleOnOpen() {
    if(this.animation && this.fade && !this.style.transition) {
      this.style.transition = `opacity var(--animation-duration, 0.2s) var(--animation-function, linear) var(--animation-delay, 0s)`;
    }

    if(this.open) {
      if(!this.fade) {
        this.style.transition = '';
      }
      this.style.visibility = 'visible';
      this.app.setAttribute('aria-hidden', 'true');
      this.setActiveStyle();
      if(
        !this.animation 
        || document.readyState === 'interactive'
        || !this.fade
      ) {
        this.focusFirst();
      }
    } else {
      if(!this.fade) {
        this.style.transition = 'opacity 0s var(--animation-delay)';
      }
      this.app.setAttribute('aria-hidden', 'false');
      this.setHideStyle();
      if(!this.animation) {
        this.focusBack();
      }
    }
    this.setTabIndex();
  }

  // TODO: Fix any type
  private handleOnTransitionEnd = (e: any) => {
    if(e.target !== this) {
      return;
    }
    if(this.open) {
      this.focusFirst();
    } else {
      this.changeModalClassList('remove', 'active');
      this.changeModalClassList('remove', 'hide');
      this.style.visibility = 'hidden';
      this.focusBack();
    }
  }

  private isFocusable(target: HTMLElement, element: HTMLElement) {
    let activeElement = null;
    if(this.node.shadowRoot) {
      activeElement = this.node.shadowRoot?.activeElement;
    } else {
      activeElement = document.activeElement;
    }
    return activeElement !== target && activeElement === element;
  }

  private focusFirstElement(target: HTMLElement, node: HTMLElement) {
    const children = node.children;
    for(let i = 0; i < children.length; i++) {
      const child = children[i] as HTMLElement;
      child.focus();
      if(this.isFocusable(target, child) || this.focusFirstElement(target, child)) {
        return true;
      }
    }
    return false;
  }

  private focusLastElement(target: HTMLElement, node: HTMLElement) {
    const children = node.children;
    for(let i = children.length - 1; i >= 0; i--) {
      const child = children[i] as HTMLElement;
      child.focus();
      if(this.isFocusable(target, child) || this.focusLastElement(target, child)) {
        return true;
      }
      return false;
    }
  }
  
  private moveFocus(e: FocusEvent,callback: (target: HTMLElement, node: ModalNode) => void) {
    const target = e.target as HTMLElement;
    let node = this.node;
    if(this.node.shadowRoot) {
      node = this.shadowNode;
    }
    callback(target, node);
  }

  private moveFocusToFirst = (e: FocusEvent) => {
    this.moveFocus(e, this.focusFirstElement.bind(this));
  }
  
  private moveFocusToLast = (e: FocusEvent) => {
    this.moveFocus(e, this.focusLastElement.bind(this));
  }

  private handleOnClickBackdrop = (e: MouseEvent) => {
    if(this.disabled) {
      return;
    }
    if(e.target === this) {
      this.open = false;
    }
  }

  private handleOnKeyup = (e: KeyboardEvent) => {
    if(this.disabled) {
      return;
    }
    const key = e.keyCode;
    if(key === 27 && this.open) {
      this.open = false;
      e.stopPropagation();
    }
  }

  private handleOnDOMContentLoaded = () => {
    this.handleOnOpen();
  }
}
